pallets/liquidity-pools-gateway/src/lib.rs

```diff
@@ -461,25 +511,103 @@ pub mod pallet {
 
 						Ok::<(), DispatchError>(())
 					}
-				}
-			})?;
+				},
+			)?;
 
 			let expected_proof_count = Self::get_expected_proof_count(&router_ids)?;
 
 			Self::execute_if_requirements_are_met(
-				proof,
+				message_hash,
 				&router_ids,
 				session_id,
 				expected_proof_count,
 				domain_address,
 			)?;
 
-			Self::deposit_event(Event::<T>::MessageRecoveryExecuted { proof, router_id });
+			Self::deposit_event(Event::<T>::MessageRecoveryExecuted {
+				message_hash,
+				router_id,
+			});
+
+			Ok(())
+		}
+
+		/// Sends a message that initiates a message recovery using the
+		/// messaging router.
+		///
+		/// Can only be called by `AdminOrigin`.
+		#[pallet::weight(T::WeightInfo::initiate_message_recovery())]
+		#[pallet::call_index(12)]
+		pub fn initiate_message_recovery(
+			origin: OriginFor<T>,
+			domain: Domain,
+			message_hash: MessageHash,
+			recovery_router: [u8; 32],
+			messaging_router: T::RouterId,
+		) -> DispatchResult {
+			T::AdminOrigin::ensure_origin(origin)?;
+
+			let message = T::Message::initiate_recovery_message(message_hash, recovery_router);
+
+			Self::send_recovery_message(domain, message, messaging_router.clone())?;
+
+			Self::deposit_event(Event::<T>::MessageRecoveryInitiated {
+				domain,
+				message_hash,
+				recovery_router,
+				messaging_router,
+			});
+
+			Ok(())
+		}
+
+		/// Sends a message that disputes a message recovery using the
+		/// messaging router.
+		///
+		/// Can only be called by `AdminOrigin`.
+		#[pallet::weight(T::WeightInfo::dispute_message_recovery())]
+		#[pallet::call_index(13)]
+		pub fn dispute_message_recovery(
+			origin: OriginFor<T>,
+			domain: Domain,
+			message_hash: MessageHash,
+			recovery_router: [u8; 32],
+			messaging_router: T::RouterId,
```

Comment:
Question. Should the `domain` field always be the domain belonging to the `messaging_router` passed?

If that is the case, we can extend `RouterId: Into<Domain>`, which is already implemented in `common/routing.rs` and save the user to write that parameter.

Reply:
If agree, we can merge this anyways and do it in a next PR

Reply:
Let's do it separately and not waste another CI run?